src/win32crypto/engine.c

```diff
@@ -0,0 +1,605 @@
+    inbuf[0].BufferType = SECBUFFER_TOKEN;
+
+    SecBufferDesc inbuf_desc = { SECBUFFER_VERSION, 2, inbuf };
+
+    if (engine->status == SEC_I_CONTINUE_NEEDED ||
+        engine->status == SEC_E_INCOMPLETE_MESSAGE) {
+        ssize_t read = engine->read_fn(engine->io,
+                                      engine->inbound + engine->inbound_len,
+                                      sizeof(engine->inbound) - engine->inbound_len);
+        if (read > 0) {
+            engine->inbound_len += read;
+            UM_LOG(VERB, "read %zd bytes of handshake data", read);
+            inbuf[0].cbBuffer = (unsigned long)engine->inbound_len;
+            inbuf[0].pvBuffer = engine->inbound;
+        } else {
+            UM_LOG(ERR, "failed to read handshake data: %zd", read);
+            engine->handshake_st = TLS_HS_ERROR;
+            return engine->handshake_st;
+        }
+    }
+
+    char in[32000];
+    if (engine->protocols) {
+        memcpy(inbuf[0].pvBuffer, engine->protocols, engine->protocols_len);
+        inbuf[0].BufferType = SECBUFFER_APPLICATION_PROTOCOLS;
+        inbuf[0].cbBuffer = engine->protocols_len;
+    } else if (engine->inbound_len > 0) {
+        memcpy(in, engine->inbound, engine->inbound_len);
+        inbuf[0].BufferType = SECBUFFER_TOKEN;
+        inbuf[0].pvBuffer = in;
+        inbuf[0].cbBuffer = engine->inbound_len;
+    }
+
+    PCtxtHandle ctx = engine->handshake_st == TLS_HS_BEFORE ? NULL : &engine->ctxt_handle;
+
+    UM_LOG(TRACE, "processing %d bytes", inbuf[0].cbBuffer);
+    SECURITY_STATUS rc = InitializeSecurityContextA(
+        &engine->cred_handle, ctx, ctx ? NULL : engine->hostname,
+        req_flags, 0, 0,
+        ctx || engine->protocols ? &inbuf_desc : NULL,
+        0,
+        ctx ? NULL : &engine->ctxt_handle,
+        &outbuf_desc, &ret_flags, NULL);
+
+    if (rc < 0) {
+      LOG_ERROR(VERB, rc, "handshake result");
+    }
+
+    engine->status = rc;
+    tlsuv__free(engine->protocols);
+    engine->protocols = NULL;
+
+    if (inbuf[1].BufferType == SECBUFFER_EXTRA) {
+        UM_LOG(VERB, "extra data in handshake buffer: %lu bytes", inbuf[1].cbBuffer);
+        size_t consumed = engine->inbound_len - inbuf[1].cbBuffer;
+        memmove(engine->inbound, engine->inbound + consumed, engine->inbound_len - consumed);
+        engine->inbound_len -= consumed;
+    } else if (rc != SEC_E_INCOMPLETE_MESSAGE) {
+        engine->inbound_len = 0;
+    }
+
+    switch (rc) {
+        case SEC_E_OK:
+            engine->handshake_st = TLS_HS_COMPLETE;
+            if (engine->cert_verify_f) {
+                rc = verify_server_cert(engine);
+            }
+            if (rc != SEC_E_OK) {
+                LOG_ERROR(ERR, rc, "failed to verify server certificate");
+                engine->handshake_st = TLS_HS_ERROR;
+                return engine->handshake_st;
+            }
+            QueryContextAttributesA(&engine->ctxt_handle, SECPKG_ATTR_STREAM_SIZES, &engine->sizes);
```

Comment:
QueryContextAttributesA states the _out_ pBuffer needs to have FreeContextBuffer called at some point. Is that done somewhere else or does that not need to happen for 'reasons'? i see a call to FreeContextBuffer below but it doesn't seem to be the same pointer?

Reply:
There are a couple other places where QueryContextAttributes is called but there's no FreeContextBuffer?